File: MicroPython/tcp_client.py
from machine import Pin, SPI
import time
import w5500 

'''
Name:demo codes for RP2350-ETH
Version V1.0
For more details of the product, please refer to "https://www.seengreat.com".
'''
print("TCP Client mode test")
# W5500 pin definition
W5500_RSTn = Pin(20, Pin.OUT)
W5500_INTn = Pin(21, Pin.IN, Pin.PULL_UP)
W5500_CSn = Pin(17, Pin.OUT)
W5500_MOSI = Pin(19)
W5500_MISO = Pin(16)
W5500_SCLK = Pin(18)
 
# config W5500
def W5500_Config(eth):
    # config gateway IP to 192.168.1.1
    dat = bytearray([192, 168, 0, 1])
    eth.Write_Bytes(w5500.GAR, dat, 4)
    gateway = f"{dat[0]}.{dat[1]}.{dat[2]}.{dat[3]}"
    print("Gateway:",gateway)
    # Set the subnet mask to 255.255.255.0
    dat = bytearray([255, 255, 255, 0])
    eth.Write_Bytes(w5500.SUBR, dat, 4)
    sub_addr = f"{dat[0]}.{dat[1]}.{dat[2]}.{dat[3]}"
    print("Sub mask:",sub_addr)

    # set MAC address to 0x48,0x53,0x00,0x57,0x55,0x00
    dat = bytearray([0x48, 0x53, 0x00, 0x57, 0x55, 0x00])
    eth.Write_Bytes(w5500.SHAR, dat, 6)
    mac_address = ':'.join(map(lambda x: '{:02x}'.format(x), dat))
    print("MAC: ",mac_address)

    # set W5500 IP to 192.168.1.101
    dat = bytearray([192, 168, 0, 101])
    eth.Write_Bytes(w5500.SIPR, dat, 4)
    ip_address = f"{dat[0]}.{dat[1]}.{dat[2]}.{dat[3]}"
    print("Client IP:",ip_address)

# detect gateway
def Detect_Gateway(eth):
    # Set a target in a different subnet IP
    dat = bytearray([192 + 1, 168 + 1, 0 + 1, 20 + 1])
    eth.Write_SOCK_4_Byte(eth.socket_num, w5500.Sn_DIPR, dat)

    # set Socket to TCP mode
    eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_MR, w5500.MR_TCP)
    # open Socket 
    eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.OPEN)

    # wait
    time.sleep_ms(5)
    # Check if the Socket status is SOCK_INIT.
    if eth.Read_SOCK_1_Byte(eth.socket_num, w5500.Sn_SR) != w5500.SOCK_INIT:
        # if not，close Socket
        eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.CLOSE)
        return w5500.FALSE

    # 发起连接
    eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.CONNECT)

    while True:
        # Check if it has timed out.
        if eth.Read_SOCK_1_Byte(eth.socket_num, w5500.Sn_IR) & w5500.IR_TIMEOUT:
            # if time out，close Socket
            eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.CLOSE)
            return w5500.FALSE
        # Check if the gateway is detected.
        elif eth.Read_SOCK_1_Byte(eth.socket_num, w5500.Sn_DHAR) != 0xff:
            # If the gateway is detected ，close Socket
            eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.CLOSE)
            break

    return w5500.TRUE

# config Socket
def Socket_Config(eth):
    # set Socket port to 5000
    eth.Write_SOCK_2_Byte(eth.socket_num, w5500.Sn_PORT, 5000)
    # Set the maximum segment size to 1460
    eth.Write_SOCK_2_Byte(eth.socket_num, w5500.Sn_MSSR, 1460)
    # Set the size of the RX buffer to 2K
    eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_RXBUF_SIZE, 0x02)
    # Set the size of the TX buffer to 2K
    eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_TXBUF_SIZE, 0x02)
    # Set the destination IP address 
    dat = bytearray([192, 168, 0, 113]) #your tcp server IP
    eth.Write_SOCK_4_Byte(eth.socket_num, w5500.Sn_DIPR, dat)
    # Set the destination port (example: 5000)
    eth.Write_SOCK_2_Byte(eth.socket_num, w5500.Sn_DPORTR, 5000)
                
if __name__ == "__main__":
    # Initialize the SPI
    spi = SPI(0, baudrate=1000000, polarity=0, phase=0, sck=W5500_SCLK, mosi=W5500_MOSI, miso=W5500_MISO)
    # Initialize W5500
    eth = w5500.W5500(spi, W5500_CSn, W5500_RSTn, W5500_INTn)
    eth.socket_num = 0
    eth.get_ver()
    # delay 100ms
    time.sleep_ms(100)
    # Detect the Ethernet connection
    while True:
        time.sleep_ms(100)
        i = eth.Read_1_Byte(w5500.PHYCFGR)
        if i & w5500.LINK:
            print("linked")
            break

    # config W5500
    W5500_Config(eth)
    # detect gateway
    if Detect_Gateway(eth):
        print("detect gateway ok")

    while True:
        # Detect the Ethernet connection
        if not (eth.Read_1_Byte(w5500.PHYCFGR) & w5500.LINK):
            print("not link")
            # close Socket 0
            eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.CLOSE)
            # Wait for the Ethernet connection to be restored.
            while not (eth.Read_1_Byte(w5500.PHYCFGR) & w5500.LINK):
                time.sleep_ms(100) 
            print("linked!")
        # read Socket 0 state
        i = eth.Read_SOCK_1_Byte(eth.socket_num, w5500.Sn_SR)
        if i == w5500.SOCK_CLOSED or i == w5500.SOCK_CLOSE_WAIT:
             print("Socket closed, reconnecting...\r\n")
             eth.Write_SOCK_1_Byte(eth.socket_num, w5500.Sn_CR, w5500.CLOSE)
             time.sleep_ms(1000)
             Socket_Config(eth) 
             if not eth.Socket_Connect(eth.socket_num):
                print("Reconnect failed, retrying...\r\n")
                time.sleep_ms(1000)
                continue
        elif i == w5500.SOCK_ESTABLISHED:
            # Send data to server
            send_data = b"Hello Server!"
            eth.Write_SOCK_Data_Buffer(eth.socket_num, send_data, len(send_data))
            print("Data sent to server:", send_data)
            
            # Receive data from server
            recv_data = eth.Read_SOCK_Data_Buffer(eth.socket_num)
            if recv_data != 0:
                print("Data received from server:", recv_data)
            
            time.sleep(1)  # Wait before sending the next data

File: vscode/tcp_client/tcp_client.c
/**
 * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include "pico/stdlib.h"


#include "w5500.h"

// W5500 pin definition
//#define W5500_RSTn 20
//#define W5500_INTn 21
//#define W5500_CSn 17
//#define W5500_MOSI 19
//#define W5500_MISO 16
//#define W5500_SCLK 18


void W5500_Config(void) {
    // config gateway IP to 192.168.0.1
    uint8_t dat[4] = {192, 168, 0, 1};
    W5500_writeBytes(GAR, dat, 4);
    printf("Gateway: %d.%d.%d.%d\r\n",dat[0],dat[1],dat[2],dat[3]);
    
    // Set the subnet mask to 255.255.255.0
    uint8_t sub_mask[4] = {255, 255, 255, 0};
    W5500_writeBytes(SUBR, sub_mask, 4);
    printf("Sub mask: %d.%d.%d.%d\r\n",sub_mask[0],sub_mask[1],sub_mask[2],sub_mask[3]);

    // set MAC address to 0x48,0x53,0x00,0x57,0x55,0x00
    uint8_t mac[6] = {0x48, 0x53, 0x00, 0x57, 0x55, 0x00};
    W5500_writeBytes(SHAR, mac, 6);
    printf("MAC: 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x\r\n",mac[0],mac[1],mac[2],mac[3],mac[4],mac[5]);

    // set W5500 IP to 192.168.0.101
    uint8_t ip[4] = {192, 168, 0, 101};
    W5500_writeBytes(SIPR, ip, 4);
    printf("TCP Client IP: %d.%d.%d.%d\r\n",ip[0],ip[1],ip[2],ip[3]);
}

uint8_t Detect_Gateway(void) {
    // Set a target in a different subnet IP
    uint8_t dat[4] = {192 + 1, 168 + 1, 0 + 1, 20 + 1};
    W5500_writeSOCK4Byte(socket_num, Sn_DIPR, dat);

    // set Socket to TCP mode
    W5500_writeSOCK1Byte(socket_num, Sn_MR, MR_TCP);
    // open Socket 
    W5500_writeSOCK1Byte(socket_num, Sn_CR, OPEN);

    // wait
    sleep_ms(5);
    // Check if the Socket status is SOCK_INIT.
    if (W5500_readSOCK1Byte(socket_num, Sn_SR) != SOCK_INIT) {
        // if not，close Socket
        W5500_writeSOCK1Byte(socket_num, Sn_CR, CLOSE);
        return FALSE;
    }

    // Initiate connection
    W5500_writeSOCK1Byte(socket_num, Sn_CR, CONNECT);

    while (true) {
        // Check if it has timed out.
        if (W5500_readSOCK1Byte(socket_num, Sn_IR) & IR_TIMEOUT) {
            // if time out，close Socket
            W5500_writeSOCK1Byte(socket_num, Sn_CR, CLOSE);
            return FALSE;
        }
        // Check if the gateway is detected.
        else if (W5500_readSOCK1Byte(socket_num, Sn_DHAR) != 0xff) {
            // If the gateway is detected ，close Socket
            W5500_writeSOCK1Byte(socket_num, Sn_CR, CLOSE);
            break;
        }
    }

    return TRUE;
}

void Socket_Config(void) {
    // set Socket 0 port to 5000
    uint16_t port = 5000;
    W5500_writeSOCK2Byte(socket_num, Sn_PORT, port);
    // Set the maximum segment size to 1460
    W5500_writeSOCK2Byte(socket_num, Sn_MSSR, 1460);
    // Set the size of the RX buffer to 2K
    W5500_writeSOCK1Byte(socket_num, Sn_RXBUF_SIZE, 0x02);
    // Set the size of the TX buffer to 2K
    W5500_writeSOCK1Byte(socket_num, Sn_TXBUF_SIZE, 0x02);
    uint8_t server_ip[4] = {192, 168, 0, 113}; // your TCP server IP
    W5500_writeSOCK4Byte(socket_num, Sn_DIPR, server_ip);
    // Set the destination port (example: 5000)
    W5500_writeSOCK2Byte(socket_num, Sn_DPORTR, 5000);
}

void Loop_Back(void) {
    uint8_t buffer[1460];
    uint16_t len = W5500_readSOCKDataBuffer(socket_num, buffer);
    if (len == 0) {
        return;
    }
    
    printf("len: %d ",len);
    for (uint16_t i = 0; i < len; i++) {
        putchar_raw(buffer[i]); 
    }
    printf("\r\n");
    
    W5500_writeSOCKDataBuffer(socket_num, buffer, len);
}

void main() {
    stdio_init_all();
    irq_set_enabled(IO_IRQ_BANK0, true); 
    printf("TCP Client mode test\r\n");
    sleep_ms(2000);
    W5500_Gpio_Init(); // Initialize W5500
    sleep_ms(2000);
    W5500_getVer();
    // Detect the Ethernet connection
    while (true) {
        sleep_ms(100);
        uint8_t i = W5500_read1Byte(PHYCFGR);
        if (i & LINK) {
            printf("linked\r\n");
            break;
        }
    }
    // config W5500
    W5500_Config();
    // detect gateway
    if (Detect_Gateway()) {
        printf("detect gateway ok\r\n");
    }
    while(1){
        // Detect the Ethernet connection
        if (!(W5500_read1Byte(PHYCFGR) & LINK)) {
            printf("not link\r\n");
            // close Socket 0
            W5500_writeSOCK1Byte(socket_num, Sn_CR, CLOSE);
            // Wait for the Ethernet connection to be restored.
            while (!(W5500_read1Byte(PHYCFGR) & LINK)) {
                sleep_ms(100); 
            }
            printf("linked!");
        }
        
        // read Socket state
        uint8_t i = W5500_readSOCK1Byte(socket_num, Sn_SR);
        if (i == SOCK_CLOSED || i == SOCK_CLOSE_WAIT) {
            printf("Socket closed, reconnecting...\r\n");
            W5500_writeSOCK1Byte(socket_num, Sn_CR, CLOSE);
            sleep_ms(100);
            Socket_Config(); 
            if (!W5500_socketConnect(socket_num)) {
                printf("Reconnect failed, retrying...\r\n");
                sleep_ms(1000);
                continue;
            }
        }
        else if (i == SOCK_ESTABLISHED) {
            // Send data to server
            const char* send_data = "Hello Server!";
            uint16_t send_len = strlen(send_data);
            uint8_t data_buffer[send_len];
            memcpy(data_buffer, send_data, send_len);
            
            W5500_writeSOCKDataBuffer(socket_num, data_buffer, send_len);
            printf("Data sent to server: ");
            for(uint16_t i = 0; i < send_len; i++)
            {
                putchar_raw(send_data[i]);
            }
            printf("\r\n");
            // Receive data from server
            uint8_t recv_buffer[1460];
            uint16_t recv_len = W5500_readSOCKDataBuffer(socket_num, recv_buffer);
            if (recv_len != 0) {
                printf("Data received from server: ");
                for (uint16_t i = 0; i < recv_len; i++) {
                    putchar_raw(recv_buffer[i]);
                }
                printf("\r\n");
            }
            sleep_ms(1000); // Wait before sending the next data
        }
    }
}
